python/ray/dag/tests/experimental/test_torch_tensor_dag.py

```diff
@@ -1792,6 +1796,84 @@ def test_torch_tensor_nccl_all_reduce_scheduling_one_ready_group(ray_start_regul
     compiled_dag.teardown()
 
 
+@pytest.mark.parametrize("ray_start_regular", [{"num_cpus": 4}], indirect=True)
+def test_torch_tensor_copy_before_nccl_all_reduce(ray_start_regular):
+    """
+    Test that a tensor is copied before all reduce is launched.
+    That is, the tensor is not all-reduced in place and can be reused
+    in the DAG as input to other nodes.
+    """
+    if not USE_GPU:
+        pytest.skip("NCCL tests require GPUs")
+
+    assert (
+        sum(node["Resources"].get("GPU", 0) for node in ray.nodes()) > 1
+    ), "This test requires at least 2 GPUs"
+
+    actor_cls = TorchTensorWorker.options(num_cpus=0, num_gpus=1)
+
+    num_workers = 2
+    workers = [actor_cls.remote() for _ in range(num_workers)]
+
+    shape = (10,)
+    dtype = torch.float16
+    with InputNode() as inp:
+        x = workers[0].send.bind(shape, dtype, inp)
+        y = workers[1].send.bind(shape, dtype, inp)
+
+        allreduce = collective.allreduce.bind([x, y])
+
+        u = workers[1].recv_tensor.bind(x)
+        v = workers[0].recv_tensor.bind(y)
+
+        dag = MultiOutputNode([*allreduce, u, v])
+
+    compiled_dag = dag.experimental_compile()
+
+    value = 10
+    ref = compiled_dag.execute(value)
+    result = ray.get(ref)
+    result = [tensor.to("cpu") for tensor in result]
+    expected_reduced_tensor_value = torch.ones(shape, dtype=dtype) * value * 2
+    expected_original_tensor_value = torch.ones(shape, dtype=dtype) * value
+    assert torch.equal(result[0], expected_reduced_tensor_value)
+    assert torch.equal(result[1], expected_reduced_tensor_value)
+    assert torch.equal(result[2], expected_original_tensor_value)
+    assert torch.equal(result[3], expected_original_tensor_value)
+
+    compiled_dag.teardown()
+
+
+@pytest.mark.parametrize("ray_start_regular", [{"num_cpus": 4}], indirect=True)
+def test_torch_tensor_nccl_all_reduce_non_tensor_input(ray_start_regular):
+    """
+    Test that an error is thrown when an all-reduce takes non-tensor input.
+    """
+    if not USE_GPU:
+        pytest.skip("NCCL tests require GPUs")
+
+    assert (
+        sum(node["Resources"].get("GPU", 0) for node in ray.nodes()) > 0
+    ), "This test requires at least 1 GPUs"
+
+    actor_cls = TorchTensorWorker.options(num_cpus=0, num_gpus=1)
+    worker = actor_cls.remote()
+
+    with InputNode() as inp:
+        nontensor = worker.send.bind((10,), torch.float16, inp, False)
```

Comment:
Nit: non_tensor.
Why is this a non-tensor? What does `send` return?

Reply:
Changed to `worker.send.bind((10,), torch.float16, inp, send_tensor=False`.
`send` returns 1 if `send_tensor` is `False`.